### ISSUE TITLE
Index ENCODE snapshot as PoC

### ISSUE THREAD

**Issue by hannes-ucsc** [DataBiosphere/azul#4224]
![AnVIL ENCODE ERM - AnVIL_ENCODE_default_v1](https://user-images.githubusercontent.com/5143256/171455336-1b512400-c375-484a-adea-4f5f71b00f6b.png)

https://lucid.app/lucidchart/ce4768f5-aaa3-444c-8fa5-133b6df0a99a/edit?viewport_loc=-3237%2C-751%2C3141%2C1773%2CWc6hws7PYHyO&invitationId=inv_fb6c8d90-b813-440e-b7ec-d63ed6a8a035#

Write a repository plugin and a metadata plugin to index the `AnVIL_ENCODE_default_v1` snapshot in the `datarepo-e8d615a1` project.

The repository plugin should be called `tdr_anvil` (or `tdranvil` if the backslash causes issues).

The metadata plugin should be called `anvil`.

The catalog should be named `anvil-encode`.

The current `tdr` plugin will at some point be renamed to `tdr_hca`. We need two plugins for TDR because the table schema of the snapshots is completely different between HCA and AnVIL.

The metadata plugin should define the following (outer and inner) entity types: `dataset` (similar to `project` in HCA), `donor` (only an inner entity in HCA), `biosample` (`sample` in HCA), `file` (same in HCA), `library` (no direct correspondence in HCA) and `activity` (only an inner entity in HCA, correspond to what HCA calls *process* aka entry in `links.json`). We'll ignore the `antibody` table for now.

Generally speaking, the columns in the snapshot tables become properties of inner entities in the Azul index. Azul distinguishes between two different forms that inner entities manifest in: The first form manifests when the inner entity type is the same as the outer entity type (`hits.contents.{entity_type}` in the response to `/index/{entity_type}`), the second form manifests when it is not. Let's refer to the first form as *pivotal*. For pivotal inner entities, include properties for the columns marked as `-`, `+`, `?` and `!` in the ERD above. For non-pivotal entities only include those marked `!` and `+`. 

The age-related columns of biosample should be indexed into an age range with unit as we do for HCA. The `donor_age_at_collection_age_stage` is the developmental stage of the donor at the time the sample was collected.

The `activity` entity type is special in that it must be polymorphic. There are just too many different `…activity` tables in the snapshot and I think polymorphism is the right approach in this case, to prevent combinatorial explosion. The set of properties in a pivotal activity entity is the union of the indexed columns of all activity tables. All properties in the union must be present in the entity. If a property doesn't correspond to a table column for a concrete activity, it should be set to `null` (`None` in Python). Add a synthetic discriminator property called `activity_type` whose value is name of the table the entity was indexed from. For non-pivotal activity entities, include only potentially useful, low-cardinality values.

When naming entity types and properties, avoid deviating from the naming conventions in the snapshot, even if the snapshot naming conventions are verbose or awkward (`donor_age_at_collection_age_unit` or `uses_sample_biosample_id`). I'll work with Broad to streamline the column names in future snapshots.

For `RepositoryPlugin.list_bundles` use following query:

```sql
SELECT
  file_id,
  FROM `datarepo-e8d615a1.AnVIL_ENCODE_default_v1.file` AS f
  LEFT JOIN `datarepo-e8d615a1.AnVIL_ENCODE_default_v1.alignmentactivity` AS ama ON f.file_id IN UNNEST(ama.uses_file_id)
  LEFT JOIN `datarepo-e8d615a1.AnVIL_ENCODE_default_v1.analysisactivity` AS asa ON f.file_id IN UNNEST(asa.derived_from_file_id)
  LEFT JOIN `datarepo-e8d615a1.AnVIL_ENCODE_default_v1.experimentactivity` AS exa ON f.file_id IN UNNEST(exa.used_file_id)
  WHERE
    STARTS_WITH(CAST(SHA1('bundles:' || f.file_id) AS STRING FORMAT 'HEX'), 'caf')
    AND ama.alignmentactivity_id IS NULL
    AND asa.analysisactivity_id IS NULL
    AND exa.experimentactivity_id IS NULL
```

where `caf` is the partition prefix. A more efficient partitioning scheme would be `MOD(ABS(FARM_FINGERPRINT(f.file_id)), 1000) = 421` but it doesn't lend itself well to the notion of partition prefix. The left joins and the `IS NULL` terms in the `WHERE` clause are there to determine the most derived files, i.e. the file leaves of the graph. We may miss leaves of other types (donors, biosamples or libraries) and we'll add them as part of another ticket.

For `RepositoryPlugin.list_partitions` use the same query but with `count(file_id)`.

For `BundleUUID`, use the first 16 bytes (32 nibbles) of the SHA-1 of the `file_id` column prefixed with `bundles:`:

```python
sha1(('bundles:'+'ENCFF957VQP').encode()).hexdigest()[:32]
'caf2ed4ebeb4e192090fd5cc8ba5f851'
```

`ENCFF957VQP` is an example `file_id`.

```sql
SELECT CAST(LEFT(SHA1('bundles:' || 'ENCFF957VQP'), 16) AS STRING FORMAT 'hex')
caf2ed4ebeb4e192090fd5cc8ba5f851
```

We can't use proper v5 UUIDs because BigQuery doesn't have support for those, unfortunately. I've raised an issue with the Broad to include UUIDs natively in as primary keys in all AnVIL snapshots.

Hard-code all versions to a fixed, constant value, say, midnight Jun 1, 2022.

For entity IDs use the same SHA-1 hash, but with the respective entity type as the prefix:

```python
sha1(('files:'+'ENCFF957VQP').encode()).hexdigest()[:32]
'4bae6247b88e2c0a7dcd58ab2e5f0248'
```

Now to the complicated part, actually extracting a subgraph:

Given a `BundleUUID`, load the corresponding file row.

```sql
SELECT 
  f.file_id 
  FROM `datarepo-e8d615a1.AnVIL_ENCODE_default_v1.file` AS f
  WHERE CAST(LEFT(SHA1('bundles:' || f.file_id), 16) AS STRING FORMAT 'hex') = 'caf2ed4ebeb4e192090fd5cc8ba5f851'
```

Next, find all activities that generate the file and their respective inputs:

```sql
WITH file AS (
  SELECT f.file_id FROM `datarepo-e8d615a1.AnVIL_ENCODE_default_v1.file` AS f 
  WHERE f.file_id IN ('ENCFF957VQP')
)
SELECT
	  'alignmentactivity' as activity_type, 
	  ama.alignmentactivity_id as activity_id, 
	  ama.uses_file_id as uses_file_id, 
	  [] as uses_biosample_id,
	  [] as library_id
  FROM file AS f
  JOIN `datarepo-e8d615a1.AnVIL_ENCODE_default_v1.alignmentactivity` AS ama 
  	ON f.file_id IN UNNEST(ama.generated_file_id)
UNION ALL SELECT
	  'analysisactivity', 
	  asa.analysisactivity_id, 
	  asa.derived_from_file_id, 
	  [],
	  []
  FROM file AS f
  JOIN `datarepo-e8d615a1.AnVIL_ENCODE_default_v1.analysisactivity` AS asa 
  	ON f.file_id IN UNNEST(asa.generated_file_id)
UNION ALL SELECT
	  'assayactivity', 
	  aya.assayactivity_id, 
	  aya.uses_sample_biosample_id, 
	  aya.uses_sample_biosample_id,
	  aya.library_id
  FROM file AS f
  JOIN `datarepo-e8d615a1.AnVIL_ENCODE_default_v1.assayactivity` AS aya 
  	ON f.file_id IN UNNEST(aya.generated_file_id)
UNION ALL SELECT
	  'sequencingactivity', 
	  sqa.sequencingactivity_id, 
	  sqa.uses_sample_biosample_id, 
	  sqa.uses_sample_biosample_id,
	  sqa.library_id
  FROM file AS f
  JOIN `datarepo-e8d615a1.AnVIL_ENCODE_default_v1.sequencingactivity` AS sqa 
  	ON f.file_id IN UNNEST(sqa.generated_file_id)
UNION ALL SELECT
	  'experimentactivity', 
	  exa.experimentactivity_id, 
	  exa.generated_file_id, 
	  exa.uses_sample_biosample_id,
	  exa.library_id
  FROM file AS f
  JOIN `datarepo-e8d615a1.AnVIL_ENCODE_default_v1.experimentactivity` AS exa 
  	ON f.file_id IN UNNEST(exa.generated_file_id)
```

Stick with these activity table aliases so its easier for me to read the queries in your code. For `file`, `biosample`, `donor` and `library` use `f`, `s`, `d` and `l` respectively.

The sole parameter to this query is a list of file IDs. The query returns one row per activity type, each row containing the type of activity, its ID along with the the IDs of the upstream files, biosamples and libraries. You'll note that we only care about the IDs here since we're only trying to determine the graph structure. The query above is the only one that requires a `UNION` — for fetching the different types of activities in a single round-trip.

Also note that there is a tail recursion at play here: the file IDs returned by the query must be fed back to the query. More on that below.

Execute analogous queries for determining the upstream entities for a given library and a given biosample. Note that the snapshot doesn't represent the activities for deriving biosamples from other biosamples. Instead, the `biosample` table has a foreign key to itself. Similarly, there is no activity table representing the collections of samples from a donor but `biosample` has a foreign key directly to `donor` instead.

The tail recursion loop must be repeated until no new entities are discovered which means that the subgraph is complete. I don't think there are any cycles in the snapshot but they should be handled gracefully. To track completeness, I would use sets of primary keys. To represent the graph, I would use an adjacency list similar to `links.json` in HCA. We must restrict ourselves to a graph representation that is JSON so that it may be canned. This means that using Python object references to represent edges is out.

Because it requires an extra query roundtrip to look up the primary key for any given entity ID (the SHA-1 of the prefixed primary key), the loop body should work on sets of primary keys as opposed to entity IDs, and primary keys should be converted to entity IDs as late as possible.

Once the subgraph is complete, fetch the indexed columns for all files in the subgraph. Try to do so in a single query (using `IN (…)` in the `WHERE` clause). I hope we don't reach size limits here. If we do, the input set of entities must be batched and one query per batch must be run. Repeat this for activities, libraries, biosamples and donors.

Hard-code the `project` entity. This snapshot does not include that information. A future snapshot will.

`RepositoryPlugin.fetch_bundle` returns an `azul.indexer.Bundle` object. You are free to choose the structure of the values of the `.manifest` and `.metadata_files` attributes as long as it can be serialized as JSON (see above). There are no DRS file IDs or content hashes in the snapshot. Hard-code mock content hashes and treat all files as phantoms.

In the interest of making the schedule, it would be acceptable to skimp on test coverage, or break non-essential functionality such as manifests and direct downloads. The goal is to demonstrate the `/index` endpoints in a personal deployment.

If the complexity balloons to a degree that the schedule is at risk, consider ignoring library and activity tables and instead rely on the redundant foreign keys between file and file and biosample and file.

Here's the simplified, internal model for this snapshot:

![AnVIL ENCODE ERM - Azul internal model(2)](https://user-images.githubusercontent.com/5143256/171455036-3928b1af-40f0-44f4-85f1-e211d91b66dc.png)

Until we're given access to an issue tracker at the Broad, please record any defects of the snapshot in [this document](https://docs.google.com/document/d/1WMI8_cuYIzKEZa-pEENsC1r5Zrf4JBY6Br6AL6iJjFc/edit?usp=sharing) *and* notify the stakeholders via email. @theathorn can tell you who they are.

**Comment by hannes-ucsc** [comment]
For demo, show that the number of hits returned by the `/files/hits` endpoint of the `anvildev` deployment approximates the number of `file` table rows in the underlying snapshot, minus adjustments for orphaned files, bad foreign keys etc.
